src/libsync/discovery.cpp

```diff
@@ -0,0 +1,1218 @@
+            // Take the things to write to the db from the "other" node (i.e: info from server).
+            // Do a lookup into the csync remote tree to get the metadata we need to restore.
+            qSwap(item->_size, item->_previousSize);
+            qSwap(item->_modtime, item->_previousModtime);
+            return false;
+        }
+        break;
+    }
+    case CSYNC_INSTRUCTION_REMOVE: {
+        const auto perms = item->_remotePerm;
+        if (perms.isNull()) {
+            // No permissions set
+            return true;
+        }
+        if (!perms.hasPermission(RemotePermissions::CanDelete)) {
+            qCWarning(lcDisco) << "checkForPermission: RESTORING" << item->_file;
+            item->_instruction = CSYNC_INSTRUCTION_NEW;
+            item->_direction = SyncFileItem::Down;
+            item->_isRestoration = true;
+            item->_errorString = tr("Not allowed to remove, restoring");
+            return true; // (we need to recurse to restore sub items)
+        }
+        break;
+    }
+    default:
+        break;
+    }
+    return true;
+}
+
+void ProcessDirectoryJob::subJobFinished()
+{
+    auto job = qobject_cast<ProcessDirectoryJob *>(sender());
+    ASSERT(job);
+
+    _childIgnored |= job->_childIgnored;
+    _childModified |= job->_childModified;
+
+    if (job->_dirItem)
+        emit _discoveryData->itemDiscovered(job->_dirItem);
+
+    int count = _runningJobs.removeAll(job);
+    ASSERT(count == 1);
+    job->deleteLater();
+    QTimer::singleShot(0, _discoveryData, &DiscoveryPhase::scheduleMoreJobs);
+}
+
+int ProcessDirectoryJob::processSubJobs(int nbJobs)
+{
+    if (_queuedJobs.empty() && _runningJobs.empty() && _pendingAsyncJobs == 0) {
+        _pendingAsyncJobs = -1; // We're finished, we don't want to emit finished again
+        if (_dirItem) {
+            if (_childModified && _dirItem->_instruction == CSYNC_INSTRUCTION_REMOVE) {
+                // re-create directory that has modified contents
+                _dirItem->_instruction = CSYNC_INSTRUCTION_NEW;
+            }
+            if (_childModified && _dirItem->_instruction == CSYNC_INSTRUCTION_TYPE_CHANGE) {
+                if (_dirItem->_direction == SyncFileItem::Up) {
+                    _dirItem->_type = ItemTypeDirectory;
+                    _dirItem->_direction = SyncFileItem::Down;
+                }
+                _dirItem->_instruction = CSYNC_INSTRUCTION_CONFLICT;
+            }
+            if (_childIgnored && _dirItem->_instruction == CSYNC_INSTRUCTION_REMOVE) {
+                // Do not remove a directory that has ignored files
+                _dirItem->_instruction = CSYNC_INSTRUCTION_NONE;
+            }
+        }
+        emit finished();
+    }
+
+    int started = 0;
+    foreach (auto *rj, _runningJobs) {
+        started += rj->processSubJobs(nbJobs - started);
+        if (started >= nbJobs)
+            return started;
+    }
+
+    while (started < nbJobs && !_queuedJobs.empty()) {
```

Comment:
Not every start()ed job will increase `_currentlyActiveJobs` by one. Maybe this should loop until the desired number for that variable is reached?

Reply:
I think it's fine as we don't want to have too many jobs in parallel.
The jobs that don't increase _currentlyActiveJobs will anyway call this function again.